### PR TITLE
Support BR on Azure Blob Storage

### LINKED ISSUE #4533: Support BR on Azure Blob Storage
## Feature Request

**Is your feature request related to a problem? Please describe:**
The TiDB is deployed in AKS and may run on any cloud or on-premises. Needs to backup to Azure blob storage using TiDB operator. Currently, BR supports this feature, but the TiDB operator does not support the feature yet.

**Describe the feature you'd like:**

- Backup works for TiDB on Azure Blob Storage.
- Restore works for TiDB on Azure Blob Storage.

**Describe alternatives you've considered:**

- Technicial spike for the entire workflow
- Extend Backup/Restore/BackupSchedule CR
- Extend Backup/Restore/BackupSchedule controller 
- Manage and GC the Azure Blob Storage with proper privileges 
- Configure the Tikv authentication mode for the Azure Blob Storage

### PR DESCRIPTION
**Pull request by xu21yingan**
### What problem does this PR solve?

Closes #4533

### What is changed and how does it work?

- adding `azblob` field in CR -`Backup`/`Restore`/`ScheduleBackup` to control backup or restore on  Azure Blob Storage.
- adding credentials for provider azblob using AD or shared key.
- adding `envFrom` field in `TidbCluster`-`TiKVSpec`/`PDSpec`/... - `ComponentSpec` for using AAD credentials on any cloud or on-premises.
- adding GC for azblob while `cleanPolicy` field in CR-`Backup` is Delete.

### Code changes

- [x] Has Go code change
- [x] Has examples related yaml change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [x] Manual test 
- [ ] No code 

Manual test (add detailed scripts or steps below):

- `Backup` CR
  - generate some data in the database
  
  - > Using shared key credentials 
    
    1. run `kubectl create secret generic azblob-secret --from-literal=AZURE_STORAGE_ACCOUNT={test_account} --from-literal=AZURE_STORAGE_KEY={test_key}`
    2. run `kubectl apply -f examples/backup/backup-azblob.yaml`
    
  - > Using AAD credentials 
 
    1. run `kubectl create secret generic azblob-secret --from-literal=AZURE_CLIENT_ID={test_cid} --from-literal=AZURE_TENANT_ID={test_tid} --from-literal=AZURE_CLIENT_SECRET={test_secret} --from-literal=AZURE_STORAGE_ACCOUNT={test_account}`
    2. run `kubectl patch tc basic --type merge -p '{"spec":{"tikv":{"envFrom":[{"secretRef":{"name":"azblob-secret"}}]}}}'`, then wait for TiKV to finish restarting
    3. look forward to setting `sendCredToTikv` value to false, before running `kubectl apply -f examples/backup/backup-azblob.yaml`
    
  - wait for the Job to complete, then the backup metadata and sst files will appear in the /test1/t1 folder on Azure Blob Storage
  - run `kubectl delete -f examples/backup/backup-azblob.yaml`  and notice that backup files are deleted because `spec.cleanPolicy` is set to Delete
 
- `Restore` CR
  - delete data from the database  
  - run `kubectl apply -f examples/backup/restore-azblob.yaml`. the location should point to the same azblob folder in the `Backup` CR     
  - wait for the Job to complete, then the metadata and sst files will be restored into the database

- `BackupSchedule` CR
  - generate some data in the database
  - Refer to `Backup` CR above in terms of using shared key or AAD credentials
  - run `kubectl apply -f examples/backup/backup-schedule-azblob.yaml`, which schedules a backup job every minute.
  - wait for a few minutes and there should be multiple backups in /test1/t1 folder on azblob
  - run `kubectl delete -f examples/backup/backup-schedule-azblob.yaml` and notice that backup files are deleted because `spec.backupTemplate.cleanPolicy` is set to Delete

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [x] Need to cherry-pick to the release branch 
- [x] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Support BR on Azure Blob Storage in TiDB Operator
```

### CODE REVIEW

File: cmd/backup-manager/app/util/remote.go [hunk 1]
```diff
@@ -405,6 +431,120 @@ func newGcsStorage(conf *gcsConfig) (*blob.Bucket, error) {
...
+
+// newAzblobStorage initialize a new azblob storage
+func newAzblobStorage(conf *azblobConfig) (*blob.Bucket, error) {
+	account := os.Getenv("AZURE_STORAGE_ACCOUNT")
+	if len(account) == 0 {
+		return nil, errors.New("No AZURE_STORAGE_ACCOUNT")
+	}
+
+	// Azure AAD Service Principal with access to the storage account.
+	clientID := os.Getenv("AZURE_CLIENT_ID")
```
Comment (KanShiori) [inline cmd/backup-manager/app/util/remote.go:456]: why not use the constant in `pkg/backup/constants`?

Comment (KanShiori) [reply cmd/backup-manager/app/util/remote.go:456]: oh, I get 😄 . It is used for secrets

File: tools/tools.go [hunk 2]
```diff
@@ -11,6 +11,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//go:build tools
```
Comment (KanShiori) [inline tools/tools.go:14]: have to remove this line

**Review by KanShiori — APPROVED** [review]

**Comment by xu21yingan** [comment]
> Hi @xu21yingan, could you please add descriptions about test cases in `Manual test`?

Ok, I will summarize and add this information as soon as possible.

**Comment by xu21yingan** [comment]
> 

done. PTAL

**Comment by xu21yingan** [comment]
> > Hi @xu21yingan, could you please add descriptions about test cases in `Manual test`?
> 
> Ok, I will summarize and add this information as soon as possible.

done. PTAL

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**